### ISSUE TITLE
Automatic uppercase-to-lowercase coercion for sloppy chording

### ISSUE THREAD

**Issue by matthew-piziak** [magit/transient#40]
I have a custom command that I bind to `C-c L m`. Sometimes I hold down the shift key too long and type `C-c L M` by accident. With `global-set-keybinding`, Emacs is forgiving: since there is no command actually _bound_ to `C-c L M` it automatically coerces it to `C-c L m`. Do you think that `transient` should have the same behavior, or should I handle it some other way? (I suppose I could also try to get better at precise inputs but that may be putting too many demands on the fallible flesh.)

**Comment by tarsius** [comment]
Could you please point me to the documentation of this feature. I had a look but could not find it. Knowing how this works for regular key bindings would speed-up adapting or re-implementing the feature for transients.

**Comment by matthew-piziak** [comment]
@tarsius Certainly. Elisp manual node `Key Sequence Input`:
>If an input character is upper-case (or has the shift modifier) and has no key binding, but its lower-case equivalent has one, then ‘read-key-sequence’ converts the character to lower case.  Note that ‘lookup-key’ does not perform case conversion in this way.

**Comment by matthew-piziak** [comment]
And I believe this is the location in source: https://github.com/emacs-mirror/emacs/blob/54b92132e1ec16565d59d6d9f8ff8910f38843b2/src/keyboard.c#L9839

**Comment by matthew-piziak** [comment]
> When `transient` decides in `transient--pre-command` that the binding is no good, then the translation actually has already occurred. In most cases there will be no translation. When the transient popup is active, then it is _not_ the current buffer, the buffer that was current before the transient was invoked is still the current buffer. In many cases the translation would not work because the untranslated key is bound. For example if you press `C` in the `transient-commit` popup, then that is bound to `magit-clone` (in `magit-mode-map` and if that weren't the case, then `self-insert-command` from the global map).

I think I see what you're saying. But in this case it's `C-c L` that runs the `-dispatch` command. Once you're in the dispatch context, then presumably Transient knows what `m` and `M` are bound to.

> Frankly I think that is a terrible binding. You could make it worse using `C-c L M-m`, i.e. using three different modifiers, but you would also make it better using only one modifier, i.e. `C-c l m` or `C-c C-l m`, or even `C-c C-l C-m`. That way only one modifier has to be pressed and released at the appropriate time.

That's a fair judgement! I would have gone with `C-c l` or `C-c C-l` but they're already bound to `org-store-link` and `org-insert-link`. (Actually I wouldn't have gone with `C-c C-l` since I follow the recommendation that `C-c C-` commands are for user-specific mode-specific commands, and `C-c` for user-specific global commands)

In any case, thanks for looking into this. Even without doing anything, the conditioning is making my fingers learn the keybindings better, so the problem will solve itself over time. :)

**Comment by tarsius** [comment]
My above description of what happens wasn't so good. Let me try again.

Transient does not actually disable any of the other active keymaps, it just puts its own keymaps into effect and gives them a higher priority than any other keymaps. So if you press `M` in your example, that most likely is not translated to `m` because `M` is bound to something, if nothing else, then to `self-insert-command`.

The reason that other keymaps are not disabled is that this makes it possible to define transient prefix commands that let the user invoke foreign commands. Magit does not currently use that, but supporting this feature is useful as proven by the Hydra package. In Hydra terminology that would be a "pink" hydra I believe.

**Comment by matthew-piziak** [comment]
> Transient does not actually disable any of the other active keymaps, it just puts its own keymaps into effect and gives them a higher priority than any other keymaps. So if you press `M` in your example, that most likely is not translated to `m` because `M` is bound to something, if nothing else, then to `self-insert-command`.

Ah, I understand now. Thank you for taking the time to explain it!
